Add service to change visibility of a group

postlund opened the pull request:
**Description:**
This pull request adds a new service that makes it possible to hide or show a specific group. Using this together with automations makes it possible to create a view with dynamic behaviour, only showing groups that e.g. are interesting during a specific time of day. Maybe only show commute sensors during weekday mornings, webcams when no one is home or a specific set of light switches when it's getting dark outside?

I'm not sure if this is the best approach, but it works. Please give feedback on this so we can find the best solution 

**Example entry for `configuration.yaml` (if applicable):**
In this example a special sensor is used that reflects certain occasions during a day (just to make the automations simpler). A command_line sensor and a small python script (below) is used to create this sensor (which really should be considered a hack).

An automation triggers when the sensor changes its state and changes visibility of a group (using a script) if it matches a specific state. So, if the sensor matches "work_morning" the group will be visible and when it changes to something else it will be hidden.

``` yaml
group:
  default_view:
    entities:
      - group.work_sensors

  # Only visible when it's time to go to work
  work_sensors:
    name: Time to go to work
    entities:
      - sensor.something1
      - sensor.something2

sensor:
  - platform: command_line
    name: Occasion
    command: "python3 occasion.py"

script:
  group_visibility:
    sequence:
      - service: group.set_visibility
        data_template:
          entity_id: "{{ entity_id }}"
          visible: "{{ is_state(cond, visible_state) }}"

automation:
  - alias: Work morning
    trigger:
      - platform: state
        entity_id: sensor.occasion
      - platform: event
        event_type: homeassistant_start
    action:
      service: script.group_visibility
      data:
        entity_id: group.work_sensors
        cond: sensor.occasion
        visible_state: 'work_morning'
```

Here is an example of the occasion.py-script used above. It basically outputs different states depending on a timespan for a given set of days. As an example, it will output _work_morning_ between 06:00-07:10 monday-friday. Please note that this is not needed, the automation can be written in any way you like (I just found this easier). Also, note that the update frequency of the command line sensor is 90 seconds, so the times specified are estimates.

``` python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from datetime import time, datetime

def mk_occasion(name, start, end, days=None):
    s = start.split(':')
    e = end.split(':')
    return {'name' : name,
            'start': time(int(s[0]), int(s[1]), int(s[2])),
            'end'  : time(int(e[0]), int(e[1]), int(e[2])),
            'days' : days}

# Matching is done from top to bottom
OCCASIONS = [
    # More specific occasions
    mk_occasion('work_morning', '06:00:00', '07:10:00', range(5)),

    # General matching
    mk_occasion('weekday', '00:00:00', '23:59:59', range(5)),
    mk_occasion('weekend', '00:00:00', '23:59:59', [5, 6])
]

def get_current_occasion(occasion_list, default_occasion='normal'):
    now = datetime.now()
    for occasion in OCCASIONS:
        if occasion['start'] <= now.time() <= occasion['end'] and \
           (occasion['days'] is None or now.weekday() in occasion['days']):
            return occasion['name']
    return default_occasion

if __name__ == '__main__':
    print(get_current_occasion(OCCASIONS))
```

If a sensor only belongs to one group and that group is hidden, that sensor will "jump" to the top of the web interface. Add the sensor to an additional group if you do not want this to happen.
**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] Tests have been added to verify that the new code works.

postlund committed:
Add service to change visibility of a group

balloob commented on homeassistant/components/group.py:
```diff
@@ -266,10 +292,20 @@ def icon(self):
         """Return the icon of the group."""
         return self._icon
 
+    @asyncio.coroutine
```
You can make this a callback and use `self.hass.loop.create_task(self.async_update_ha_state())` instead of the `yield`

balloob commented on homeassistant/components/group.py:
```diff
@@ -162,6 +175,18 @@ def reload_service_handler(service_call):
+        visible = service.data.get(ATTR_VISIBLE)
+        for group in component.async_extract_from_service(
+                service, expand_group=False):
+            hass.loop.create_task(group.async_set_visible(visible))
```
You can call it directly once you made `async_set_visible` a callback.

balloob reviewed (commented):
Very nice! Just a few async comments.

balloob commented:
Oh and it seems like you have two linting errors.

postlund commented:
Thanks for the comments! :) I have updated according to your comments, fixed the lint errors and added a test that I missed. So I hope everything is fine now.

balloob commented:
Ok to merge when merge conflicts resolved 🐬

postlund commented:
Should be fixed as soon as Travis is done!

postlund commented:
I will have to study the errors a bit further...

postlund commented:
All is good and can be merged 👍

pvizeli commented:
LGTM. Please update also the documentation.

postlund commented:
I'm working on the documentation, will send a PR ASAP.

thefrenchmatt commented:
tl;dr : a noob's journey to configuring HA, from "command line sensor"'s hell and back with shell scripting.

@postlund TYSM for being so altruistic and taking the time to add it to the wiki : you made me revisit how I was configuring things in HA and your example actually gave me the final push to start learning a bit of python (I had given up about a year ago) to add OptionParser as I wanted to have several 'occasions' in the same script. While doing this, I noticed one thing that may be worth being mentioned in the wiki: **European users will want to replace now.weekday() by now.isoweekday() to have [1] for Mondays and [7] for Sundays.**

I also think the cli sensors, having hit API limits a few times, can be each adjusted by adding a **"scan_interval:" option (in seconds).**
I sometimes noticed gaps / missing data points in my graphs, and then timeouts in the logs.
I searched and couldn't see anyone with the same problem and a fix, and it didn't make sense to raise as a bug but I had a strong feeling I was responsible for creating race conditions...

I also noticed, about 6 months ago and after learning about yamllint from the wiki, that I had a serious lack of understanding quoting & escaping.
I already gave up on python so learning all these differences between script languages and programs? no thank you.
It felt like an impass but, spending more of my time in CLI, I also started trying other shells.
I now really like fish and zsh, in the last 2 months I learned about oh-my-(zsh|fish) and so finally set up everything to have them cohabit (aliases & functions, shared history, keybindings).
I still need to learn screen/tmux, just not there yet.

So basic stuff was taking less time, but I did find my old daemons (pun intended?) in variables and arrays.
This seemed even more complex than HA and YAML, but I needed solutions : new CLI sensors, simplifying my config, some components that kept breaking. Also, after running HA on different platforms (pi, desktop, docker, venv on synology, dd-wrt router), it felt like STDOUT or MQTT would not be enough, especially if I wanted to split instances.

That's how I decided to focus on text files as outputs.

I started by taking out "slow" scripts and programs (that already had several option parameters), either adapting or creating a new script, which allowed me to change the shell if needed.
I solved a lot of parsing headaches through different combinations (printf / awk / sed / wc / tr / jsonawk / jshon).
I added tests to be sure the exit codes != 0 would not lead to overwriting the previous sensor's input.
I also added a script to go through those outputs in order to have a compact timedelta (ie: starting short with +1s to a more complex +1y 1m 1d H:M:S).

After all this, one command was enough to generate more inputs, and so I turned to cron (http://crontab.guru/ was a handy tool to get the syntax right).
Then I started replacing a lot of CLI sensors with a 'cat' for each output.
It's probably why the wiki mentions HA's "command line sensor" can make a big difference, but should add a warning about its many pitfalls :)

You helped me faster manage the dynamic hiding of an ever growing number of elements (currently >350 states listeners), and I really didn't expect a ready-made solution like the one you offered, so this is my n00bs attempt in giving back, giving structure to some modest learnings.

M
